### PR TITLE
Updates for asv suite

### PR DESCRIPTION
**Pull request by qwhelan**
@jorisvandenbossche As you requested in #10849, here's the current state of affairs. The changes are:
- First patch:
  - Fix the translation script so that functions defined in the vbench `setup` string get defined as classmethods and called appropriately. This fixes most failing tests.
  - Some backwards-compatibility fixes around imports. Raising `NotImplementedError` inside `setup()` is equivalent to `SkipTest`, so I've tried to do that in a couple necessary cases (such as things using `test_parallel`)
  - Speaking of `test_parallel`, this required a bit of hackery to support not-existing. I implemented a no-op decorator by that name so that the actual definition of test cases wouldn't fail. We simply raise `NotImplementedError` rather than test without it, however.
- Second patch:
  - Support Python 3, the changes are essentially what I mentioned elsewhere.
  - `from pandas_vb_common import *` -> `from .pandas_vb_common import *`
  - Replace `xrange` with `range` and add `from pandas.compat import range` in `pandas_vb_common`
  - Replace `string.uppercase` with `string.ascii_uppercase`
- Third patch
  - `asv` instructions in `contributing.rst`, largely lifted from the vbench description. The current examples are minimalist, but most comparable to what vbench outputs.
- Fourth patch:
  - Re-run the translation script. I think some other additions may be included in the first patch due to having run it a few times while making changes.

cc @jreback @TomAugspurger

### CODE REVIEW

File: asv_bench/benchmarks/packers.py [hunk 1]
```diff
@@ -575,13 +569,19 @@ def remove(f):
...
     def time_packers_write_json_date_index(self):
         self.df.to_json(self.f, orient='split')
 
     def teardown(self):
-        remove(self.f)
+        self.remove(self.f)
+
+    def remove(self, f):
+        try:
+            os.remove(self.f)
```
Comment (jreback) [inline asv_bench/benchmarks/packers.py]: I would put all of these in a base class, maybe ``Base`` and inherit all tests from it (rather than ``object``). You might want several levels of this for various ones (e.g. ``IOBase``) or whatever. Just to remove some boilerplate.

File: asv_bench/benchmarks/strings.py [hunk 2]
```diff
@@ -116,278 +116,278 @@ class strings_endswith(object):
...
-        self.many = make_series(('matchthis' + string.uppercase), strlen=19, size=10000)
-        self.few = make_series(('matchthis' + (string.uppercase * 42)), strlen=19, size=10000)
+        self.many = self.make_series(('matchthis' + string.ascii_uppercase), strlen=19, size=10000)
+        self.few = self.make_series(('matchthis' + (string.ascii_uppercase * 42)), strlen=19, size=10000)
 
     def time_strings_repeat(self):
         self.many.str.repeat(list(IT.islice(IT.cycle(range(1, 4)), len(self.many))))
 
+    def make_series(self, letters, strlen, size):
+        return Series([str(x) for x in np.fromiter(IT.cycle(letters), count=(size * strlen), dtype='|S1').view('|S{}'.format(strlen))])
```
Comment (jreback) [inline asv_bench/benchmarks/strings.py]: e.g. these should be in the base class as well.

File: doc/source/contributing.rst [hunk 3]
```diff
@@ -512,9 +512,48 @@ entire suite.  This is done using one of the following constructs:
...
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Performance matters and it is worth considering that your code has not introduced
+performance regressions.  *pandas* is in the process of migrating to the
+`asv library <https://github.com/spacetelescope/asv>`__
+to enable easy monitoring of the performance of critical *pandas* operations.
+These benchmarks are all found in the ``pandas/asv_bench`` directory.  *asv*
+supports both python2 and python3. 
+
+.. note::
+    The *asv* benchmark suite was translated from the previous framework, vbench,
```
Comment (jreback) [inline doc/source/contributing.rst]: blank line here

File: vb_suite/io_bench.py [hunk 4]
```diff
@@ -1,8 +1,11 @@
 from vbench.api import Benchmark
 from datetime import datetime
 
-common_setup = """from pandas_vb_common import *
-from StringIO import StringIO
+common_setup = """from .pandas_vb_common import *
+try:
+    from StringIO import StringIO
+except ImportError:
```
Comment (jreback) [inline vb_suite/io_bench.py]: might as well do this in pandas_vb_common

**Comment by jorisvandenbossche** [comment]
@jreback regarding your comments, I think that mainly are the things that will have to be handled manually (@qwhelan is that correct?) That is the clean-up mentioned in #10849

**Comment by jreback** [comment]
sure that is fine
yes iirc that is a further cleanup

ok so merge when ready then

File: vb_suite/io_bench.py (hunk 4 above)
Comment (kawochen) [inline vb_suite/io_bench.py]: see #10470 -- you can probably just do `from io import StringIO` without the try-except.

File: asv_bench/benchmarks/packers.py (hunk 1 above)
Comment (qwhelan) [inline asv_bench/benchmarks/packers.py]: My expectation is that any of the benchmark modules will have a handful of classes post-cleanup. ```asv``` doesn't require different classes per benchmark, so we can put ```time_foo()``` and ```time_bar()``` on the same class. Also, these changes are all auto-generated.

File: vb_suite/io_bench.py (hunk 4 above)
Comment (qwhelan) [inline vb_suite/io_bench.py]: @kawochen Cool, wasn't aware of that. I'll change that in a minute.

**Comment by qwhelan** [comment]
@jorisvandenbossche Yes, that is correct. The only manual changes were in ```vbench_to_asv.py``` and inside ```vb_suite``` (which were propagated to ```asv_bench``` by rerunning the former). I also manually ensured some new ```asv``` tests added by others didn't get overwritten.

Just pushed some small changes (StringIO change and added a new line in contributing). I think it's good to merge.

**Comment by jorisvandenbossche** [comment]
@qwhelan Thanks a lot for this!

**Comment by qwhelan** [comment]
Of course.

I also have a bunch of asv run data I should throw up on a github.io page
so the EuroScipy folks can look into some of the regressions (won't get to
this for probably 6+ hours).

**Comment by qwhelan** [comment]
@jorisvandenbossche My run data should be accessible here: http://qwhelan.github.io/pandas_asv/ .

An explicit list of regressions can be seen here: http://qwhelan.github.io/pandas_asv/#/regressions

File: asv_bench/asv.conf.json [hunk 5]
```diff
@@ -41,7 +41,10 @@
         "sqlalchemy": [],
         "scipy": [],
         "numexpr": [],
-        "pytables": [],
+        "tables": [],
```
Comment (RafalSkolasinski) [inline asv_bench/asv.conf.json]: Why did you change ``pytables`` to ``tables``? It doesn't work for me know on ``conda`` environment. It ask if I meant ``pytables``.

Comment (qwhelan) [inline asv_bench/asv.conf.json]: @RafalSkolasinski Looks like Anaconda changed the package name - it's on PyPI as ```tables```: https://pypi.python.org/pypi/tables . I didn't want to force people to use ```conda``` or ```virtualenv```, but I don't see an easy way to support both due to this name change.

Comment (RafalSkolasinski) [inline asv_bench/asv.conf.json]: @qwhelan I see... That's weird that package is named ``tables`` in pypi. even original webpage reads http://www.pytables.org/

Comment (jorisvandenbossche) [inline asv_bench/asv.conf.json]: but the 'import name' is tables I think.

In any case, we are probably not going to be able to specify a fully correct asv.conf.json file (eg also for windows an extra entry has to be added). So I would propose to choose one case (I would say: conda on linux) and make it work for that, and then clearly explain in the docs or readme what has to be changed to use it in the other cases.

**Comment by behzadnouri** [comment]
`from pandas_vb_common import *` -> `from .pandas_vb_common import *`

has broken `test_perf.sh`. running benchmarks fails, with `ImportError: No module named pandas_vb_common`. (on python 2)

**Comment by jorisvandenbossche** [comment]
Now we did the conversion (and it is the idea we shouldn't have to do this again, as we start changing the asv benchmarks direclty), we can change this (``from pandas_vb_common import *``) back to how it was in the vb_suite benchmarks.
